langserver/diagnostics/diagnostics.go

```diff
@@ -0,0 +1,91 @@
+package diagnostics
+
+import (
+	"context"
+	"sync"
+
+	"github.com/creachadair/jrpc2"
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/hcl/v2/hclparse"
+	ilsp "github.com/hashicorp/terraform-ls/internal/lsp"
+	"github.com/sourcegraph/go-lsp"
+)
+
+type documentContext struct {
+	ctx  context.Context
+	uri  lsp.DocumentURI
+	text []byte
+}
+
+type Notifier struct {
+	sessCtx          context.Context
+	hclDocs          chan<- documentContext
+	closeHclDocsOnce sync.Once
+}
+
+func NewNotifier(sessCtx context.Context) *Notifier {
+	hclDocs := make(chan documentContext, 10)
+	go hclDiags(hclDocs)
```

Comment:
Is there a reason we need the asynchronicity?

I mean each request is already being processed in a dedicated goroutine and parallelism is controlled in the jRPC library and that's limited to 1 for ordering reasons, so we'd never actually process more than 1 at a time anyway and even if for some reason we do in the future, then I reckon the performance/parallelism would be handled by the jRPC library on handler level?

Reply:
Are push requests limited by the parallelism? Seems like this is just pushing each chan value.

Reply:
The idea is we have a diagnostics worker thread that takes in document requests and pushes diags to the client as soon as possible, but it does not ever block or add milliseconds to the `didOpen` and `didChange` handlers (and upcoming `didSave` for validation).